academic_observatory_workflows/workflows/crossref_events_telescope.py

```diff
@@ -12,383 +12,859 @@
+            dataset_id=self.api_dataset_id,
+            dag_run_id=release.run_id,
+            changefile_start_date=release.start_date,
+            changefile_end_date=release.end_date,
+        )
+        api = make_observatory_api(observatory_api_conn_id=self.observatory_api_conn_id)
+        api.post_dataset_release(dataset_release)
+
+    def cleanup(self, release: CrossrefEventsRelease, **kwargs) -> None:
+        """Delete all files, folders and XComs associated with this release."""
+
+        cleanup(dag_id=self.dag_id, execution_date=kwargs["logical_date"], workflow_folder=release.workflow_folder)
 
-    def transform(self, release: CrossrefEventsRelease, **kwargs):
-        """Task to transform the CrossrefEventsRelease release.
 
-        :param release: a CrossrefEventsRelease instance.
-        :param kwargs: The context passed from the PythonOperator.
-        :return: None.
+def parse_release_msg(msg: Dict) -> Tuple[pendulum.DateTime, pendulum.DateTime, bool, pendulum.DateTime]:
+    start_date = pendulum.parse(msg["start_date"])
+    end_date = pendulum.parse(msg["end_date"])
+    is_first_run = msg["is_first_run"]
+    prev_end_date = pendulum.parse(msg["prev_end_date"])
+
+    return start_date, end_date, is_first_run, prev_end_date
+
+
+def get_event_date(event) -> pendulum.DateTime:
+    if "updated_date" in event:
+        return pendulum.parse(event["updated_date"])
+    else:
+        return pendulum.parse(event["timestamp"])
+
+
+class Action:
+    create = "created"
+    edit = "edited"
+    delete = "deleted"
+
+
+class DayRequest:
+    def __init__(self, day: pendulum.DateTime, mailto: str):
+        self.date = day
+        self.mailto = mailto
+        self.created = EventRequest(Action.create, day, mailto)
+        self.edited = EventRequest(Action.edit, day, mailto)
+        self.deleted = EventRequest(Action.delete, day, mailto)
+
+
+def make_crossref_events_url(
+    *, action: str, start_date: pendulum.DateTime, end_date: pendulum.DateTime, mailto: str, rows: int, cursor: str
+):
+    assert action in {
+        "created",
+        "edited",
+        "deleted",
+    }, f"make_crossref_events_url: unknown action={action}, must be one of created, edited or deleted"
+
+    # Set query params and set path in URL
+    url = CROSSREF_EVENTS_HOST
+    query_params = dict()
+
+    if action == Action.create:
+        query_params["from-collected-date"] = start_date.format(DATE_FORMAT)
+        query_params["until-collected-date"] = start_date.format(DATE_FORMAT)
+    else:
+        url += f"/{action}"
+        query_params["from-updated-date"] = end_date.format(DATE_FORMAT)
+        query_params["until-updated-date"] = end_date.format(DATE_FORMAT)
+
+    if mailto is not None:
+        query_params["mailto"] = mailto
+    if rows is not None:
+        query_params["rows"] = rows
+    if cursor is not None:
+        query_params["cursor"] = cursor
+
+    req = requests.models.PreparedRequest()
```

Comment:
This looks great, I didn't know the requests library had something like this. Will a request made in this manner always work (ie. correctly format the url) for a REST API?

Reply:
Thanks. Yeah it should do, we could use this in other places.